clippy_lints/src/default_box_assignments.rs

```diff
@@ -0,0 +1,131 @@
+    ///
+    /// ### Why is this bad?
+    /// This incurs an extra heap allocation compared to assigning the boxed
+    /// storage.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let mut b = Box::new(1u32);
+    /// b = Default::default();
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let mut b = Box::new(1u32);
+    /// *b = Default::default();
+    /// ```
+    #[clippy::version = "1.89.0"]
+    pub DEFAULT_BOX_ASSIGNMENTS,
+    perf,
+    "assigning `Default::default()` to `Box<T>` is inefficient"
+}
+declare_lint_pass!(DefaultBoxAssignments => [DEFAULT_BOX_ASSIGNMENTS]);
+
+impl LateLintPass<'_> for DefaultBoxAssignments {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &'_ Expr<'_>) {
+        if let ExprKind::Assign(lhs, rhs, _) = &expr.kind {
+            let lhs_ty = cx.typeck_results().expr_ty(lhs);
+
+            // No diagnostic for late-initialized locals
+            if let Some(local) = path_to_local(lhs)
+                && !local_is_initialized(cx, local)
+            {
+                return;
+            }
+
+            let Some(inner_ty) = get_box_inner_type(cx, lhs_ty) else {
+                return;
+            };
+
+            let Some(default_trait_id) = cx.tcx.get_diagnostic_item(sym::Default) else {
+                return;
+            };
+
+            if implements_trait(cx, inner_ty, default_trait_id, &[])
+                && is_default_call(cx, rhs)
+                && !rhs.span.from_expansion()
+            {
+                span_lint_and_then(
+                    cx,
+                    DEFAULT_BOX_ASSIGNMENTS,
+                    expr.span,
+                    "creating a new box with default content",
+                    |diag| {
+                        let mut app = Applicability::MachineApplicable;
+                        let suggestion = format!(
+                            "{} = Default::default()",
+                            Sugg::hir_with_applicability(cx, lhs, "_", &mut app).deref()
+                        );
+
+                        diag.note("this creates a needless allocation").span_suggestion(
+                            expr.span,
+                            "replace existing content with default instead",
+                            suggestion,
+                            app,
+                        );
+                    },
+                );
+            }
+
+            if inner_ty.is_sized(cx.tcx, cx.typing_env())
+                && let Some(rhs_inner) = get_new_call_value(cx, rhs)
+            {
+                span_lint_and_then(cx, DEFAULT_BOX_ASSIGNMENTS, expr.span, "creating a new box", |diag| {
+                    let mut app = Applicability::MachineApplicable;
+                    let suggestion = format!(
+                        "{} = {}",
+                        Sugg::hir_with_applicability(cx, lhs, "_", &mut app).deref(),
+                        Sugg::hir_with_applicability(cx, rhs_inner, "_", &mut app),
```

Comment:
Here, `rhs_inner` may come from a macro. This is perfectly fine though, but here you will get the macro content if you use `Sugg::hir_with_applicability()`. You need to go back to the call site (which will be the root context here since an assignment coming from expansion will be filtered out) by using `Sugg::hir_with_context()`, using the `expr.span.ctxt()` context for example.

You can add a test like that:

```rust
macro_rules! mac {
    (three) => { 3u32 };
}

b = Box::new(mac!(three));
```

and check that the `.fixed` file contains `*b = mac!(three)` (and not `*b = 3u32` as it will probably right now)